### PR TITLE
[PyTorch] Schedule delayed-scaling updates after backward

### PR DESCRIPTION
**Pull request by pggPL**
# Description

Delayed-scaling (amax/scale) updates are currently owned by the first FP8 module seen in forward and executed from that module's backward. This depends on backward traversal order and breaks with unused branches, reentrant checkpointing, or schedules that delay weight-gradient computation.

This PR moves the update to the end of the autograd task instead. Each participating module requests an update and TE queues a single callback on the enclosing `GraphTask`, so the update runs once after the whole backward has finished. Reentrant TE checkpointing registers the callback on the outer task before entering its nested backward.

Plain `.backward()` needs no changes. A new `quantization_backward_scope()` lets applications define a larger logical backward, so the update runs once when the outermost scope exits:

```python
with te.quantization_backward_scope():
    for loss in microbatch_losses:  # e.g. a 1F1B pipeline schedule
        loss.backward()
```

The update always runs on scope exit, even on ranks that ran no quantized backward inside it, so every rank joins the amax reduction. Like `autocast`, the scope must be entered and exited on all ranks.

## New API

`transformer_engine.pytorch.quantization_backward_scope()` (context manager, documented in the PyTorch API reference and on the delayed scaling feature page):

- Delays the delayed-scaling update (amax reduction across the reduction group and scale recomputation for the gradient quantizers) until the outermost scope exits. Nested scopes are no-ops. A scope entered inside a running backward defers the update to the end of that backward.
- Always runs the update on exit, even if no quantized module ran backward inside the scope. Like `autocast`, it must be entered and exited on all ranks.
- Has no effect for recipes without delayed-scaling state.

Internal hooks used by the modules: `FP8GlobalStateManager.backward_quantization_update_needed()` (decided in forward, stored as a bool on the autograd ctx) and `FP8GlobalStateManager.request_backward_quantization_update()` (called from backward, queues one callback per autograd task).

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

- Queue one delayed-scaling update at the end of each autograd task instead of in the first module's backward.
- Decide in forward whether a module must request the update (`backward_quantization_update_needed()`), carry only a bool through the autograd ctx, and request from backward with `request_backward_quantization_update()`.
- Handle reentrant TE checkpoints by registering on the outer `GraphTask`.
- Add `quantization_backward_scope()` for custom logical-backward boundaries; it always updates on exit.
- Keep the existing `FP8GlobalStateManager.reduce_and_update_fp8_tensors()` implementation and CUDA graph update path.
- Add tests for plain, reentrant, non-reentrant, nested, branched, unused-branch, multi-backward, and delayed-wgrad cases, plus a distributed test where some ranks skip backward.

# Checklist:

- [x] I have read and followed the contributing guidelines
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Validation

- Black 24.4.2, Python syntax checks, and `git diff --check` pass for the scheduler-only revision.
- AST comparison confirms the existing update implementation and `RecipeState` classes match the merged `main`.
- GPU tests have not been rerun locally for this revision; CI results are pending.

### CODE REVIEW

**Comment by pggPL** [comment]
Hi @ksivaman ,

This PR changes the logic of delayed amax update - instead of relying on first in forward/last in backward module, we use PyTorch callback at the end of backward. This has multiple benefits:

- It removes the assumption that the first module in forward is the last to finish backward, which is fragile for branched computation graphs.
- It avoids losing the update when the module that would own it belongs to a branch that does not participate in backward.
- It simplifies coordination with reentrant TE checkpointing by deferring the update until the enclosing backward finishes.
- The new quantization_backward_scope() allows multiple backward calls, such as pipeline microbatches, to share one amax reduction and scale update.
- The scope also ensures that ranks which skip quantized backward still participate in the reduction, provided all ranks enter and exit the scope consistently.

The existing amax reduction and scale-update algorithm remains unchanged; this PR changes its scheduling.

**Comment by ksivaman** [comment]
@pggPL Is this tested with cuda graphs and/or pipeline parallelism configs?

We'll have to change the `make_graphed_callables` functionality to accommodate the new design.
